Core/include/Acts/Utilities/Delegate.hpp

```diff
@@ -0,0 +1,91 @@
+// This file is part of the Acts project.
+//
+// Copyright (C) 2021 CERN for the benefit of the Acts project
+//
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+#pragma once
+
+#include <functional>
+
+namespace Acts {
+
+template <typename>
+class Delegate;
+
+/// Delegate type that allows type erasure of a callable without allocation and
+/// with a single level of indirection.
+/// This type can support:
+/// - a free function pointer
+/// - a pointer to a member function alongside an instance pointer
+/// @note @c Delegate does not assume ownership of the instance.
+///          You need to ensure that the lifetime of the callable
+///          instance is longer than that of the @c Delegate.
+/// @note Currently @c Delegate only supports callables that are ``const``
+/// @tparam R Return type of the function signature
+/// @tparam Args Types of the arguments of the function signatures
+///
+template <typename R, typename... Args>
+class Delegate<R(Args...)> {
+  /// Alias of the return type
+  using return_type = R;
+  /// Alias to the function pointer type this class will store
+  using function_type = return_type (*)(const void*, Args...);
+
+ public:
+  Delegate() = default;
+
+  /// Connect a free function pointer.
+  /// @note The function pointer must be ``constexpr`` for @c Delegate to accept it
+  /// @tparam Callable The compile-time free function pointer
+  template <auto Callable>
+  void connect() {
+    m_payload = nullptr;
+    m_function = [](const void* /*payload*/, Args... args) -> return_type {
+      return std::invoke(Callable, std::forward<Args>(args)...);
+    };
```

Comment:
Out of curiosity, doesn't this defeat the purpose of reducing the number of indirections? I imagine that if you call the delegate object, you're retrieving the function pointer to this lambda first, then the lambda invokes the callable, isn't that two levels of indirections? Or am I misunderstanding your design?

Reply:
The other function pointer is a compile time function pointer. You're right that it's not **guaranteed** to be reduced to a single one, but I would think in almost all cases it should be.

Reply:
I could probably double check this in godbolt though.

Reply:
So my [godbolt check](https://godbolt.org/z/zxb5653z8) indicates that giving it the function pointer at compile-time should almost make sure the extra lambda is removed;